demo_keyframe_bundle_adjustment_meta/apps/main_program.cpp

```diff
@@ -0,0 +1,178 @@
+
+        odomGT.header.stamp = ros::Time().fromSec(timestamp);
+        odomGT.pose.pose.orientation.x = q.x();
+        odomGT.pose.pose.orientation.y = q.y();
+        odomGT.pose.pose.orientation.z = q.z();
+        odomGT.pose.pose.orientation.w = q.w();
+        odomGT.pose.pose.position.x = t(0);
+        odomGT.pose.pose.position.y = t(1);
+        odomGT.pose.pose.position.z = t(2);
+        pubOdomGT.publish(odomGT);
+
+        geometry_msgs::PoseStamped poseGT;
+        poseGT.header = odomGT.header;
+        poseGT.pose = odomGT.pose.pose;
+        pathGT.header.stamp = odomGT.header.stamp;
+        pathGT.poses.push_back(poseGT);
+        pubPathGT.publish(pathGT);
+
+        // read lidar point cloud
+        std::stringstream lidar_data_path;
+        lidar_data_path << dataset_folder << "sequences/" + sequence_number + "/velodyne/" 
+                        << std::setfill('0') << std::setw(6) << line_num << ".bin";
+        std::vector<float> lidar_data = read_lidar_data(lidar_data_path.str());
+        std::cout << "totally " << lidar_data.size() / 4.0 << " points in this lidar frame \n";
+
+        std::vector<Eigen::Vector3d> lidar_points;
+        std::vector<float> lidar_intensities;
+        pcl::PointCloud<pcl::PointXYZI> laser_cloud;
+        for (std::size_t i = 0; i < lidar_data.size(); i += 4)
+        {
+            lidar_points.emplace_back(lidar_data[i], lidar_data[i+1], lidar_data[i+2]);
+            lidar_intensities.push_back(lidar_data[i+3]);
+
+            pcl::PointXYZI point;
+            point.x = lidar_data[i];
+            point.y = lidar_data[i + 1];
+            point.z = lidar_data[i + 2];
+            point.intensity = lidar_data[i + 3];
+            laser_cloud.push_back(point);
+        }
+
+        sensor_msgs::PointCloud2 laser_cloud_msg;
+        pcl::toROSMsg(laser_cloud, laser_cloud_msg);
+        laser_cloud_msg.header.stamp = ros::Time().fromSec(timestamp);
+        laser_cloud_msg.header.frame_id = "/camera_init";
+        pub_laser_cloud.publish(laser_cloud_msg);
+
+        sensor_msgs::ImagePtr image_left_msg = cv_bridge::CvImage(laser_cloud_msg.header, "mono8", left_image).toImageMsg();
+        sensor_msgs::ImagePtr image_right_msg = cv_bridge::CvImage(laser_cloud_msg.header, "mono8", right_image).toImageMsg();
+        pub_image_left.publish(image_left_msg);
+        pub_image_right.publish(image_right_msg);
+
+        if (to_bag)
+        {
+            bag_out.write("/image_left", ros::Time::now(), image_left_msg);
+            bag_out.write("/image_right", ros::Time::now(), image_right_msg);
+            bag_out.write("/kitti/velo/pointcloud", ros::Time::now(), laser_cloud_msg);
+            bag_out.write("/path_gt", ros::Time::now(), pathGT);
+            bag_out.write("/odometry_gt", ros::Time::now(), odomGT);
+        }
+
+        line_num ++;
+        r.sleep();
+    }
+    bag_out.close();
+    std::cout << "Done \n";
+
+
+    return 0;
+}
```

Comment:
Great! I would but the nin trivial parsing part (pointlcoud, odomGt) into functions to make it more readable

Reply:
Made them into functions :)